### PR TITLE
fix(inductor): enable mid-stick torch.cat by sharing one mutation buffer layout

### LINKED ISSUE #1464: Identity handling for values smaller than a stick along the stick dimension
### What feature would you like to be added?

We would like to add a new concatenation mechanism that can handle values smaller than a stick (64 fp16 elements) along the stick dimension. For example, concatenating tensors of shape [2, 32] and [2, 32] along dim = -1.

### Why is this feature needed?

This is required for `overwrite`, which is used by `torch.cat`, `torch.pad`, and `torch.stack`.

We are seeing the occurrence of this pattern:
- `torch.cat` in Mistral-Small-3-2-24B-Instruct-2506 and gpt-oss-20b 
- `torch.stack` in Mistral-Small-3-2-24B-Instruct-2506
- `torch.pad` in granite-4.0 ( #1310 )

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

This feature is required in Q2 #940

### PR DESCRIPTION
**Pull request by yoheiueda**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

`torch.cat` whose split point is not a stick multiple previously returned wrong results silently. Such a concat lowers to **two slice writes into one freshly cloned buffer**, both landing in the *same* 64-wide stick — `[..., :32]` (offset-free but sub-stick) and `[..., 32:64]` (offset). Written natively, each pads its slice to a full stick and zeroes the tail, so whichever lands second clobbers the other.

The two writes already shared the buffer; what they did not share was a *layout*. This PR makes them agree on **one alternate layout** for that buffer, chosen once per mutation target, which moves the offset off the stick dim so both writes lower natively and the whole stick is covered.

The underlying capability is more general: offset and sub-stick slice writes into an **internal, compiler-allocated buffer** now compile at all.

```python
base[..., 32:64].copy_(src)   # into a buffer the compiler allocated, not an argument
```

---

### High-level overview

A mutation write needs its target's device layout to place the write inside a stick. When the write is offset (`[..., 32:64]`) or covers only part of a full-stick dim (`[..., :32]`), the target's natural layout cannot express it. The fix relocates the target onto an **alternate layout** whose stick dim is a different host dim, so the offset moves off the stick axis and every write becomes whole-stick.

The choice of alternate layout must be **per buffer, not per write** — aliasing siblings from one `cat` write into the same buffer and must agree. The work spans two passes that run in sequence.

**1. `propagate_layouts.py`** — decides the layout. One up-front scan picks a single alt per mutation target, so the main loop finds the decision already made instead of discovering a conflict per op:

```
propagate_spyre_tensor_layouts              # the pass; one walk over all operations
├── _scan_mutation_layout_inputs            # ADDED: up-front, one alt per target
│   ├── group aliasing writes by target buffer name
│   └── _find_alt_target_stl                #   first offset-free candidate becomes the shared alt
│       └── _is_substick_write              #   ADDED: offset-free yet shorter than a full stick?
└── for op in operations:                   # the mutation block
    ├── mutation_alts.get(target_name)      #   reuse the alt the scan committed
    ├── _align_single_source_producer        # ADDED: put the producer on the same alt so the
    │                                        #   copy edge is layout-identical and fuses
    └── op._restickify_plan = (...)          # hands the target + alt to insert_restickify
```

The scan resolves a group by target kind. An **internal buffer** takes the first alt found and forces the rest onto it, which the retarget below makes physically true for every writer. A **graph input** is caller-visible, so each of its writes is checked against the recorded alt and a disagreement raises `Unsupported` rather than being silently forced; chaining relayouts between conflicting writes is left as a `TODO`. Targets with no device layout yet — a `SpyreEmptyFallback` — are excluded.

`_find_alt_target_stl` gained an `in_arg` parameter so it prefers a candidate *reachable* from the write's input stick — one an ordinary stick-permutation restickify can actually produce — rather than letting a degenerate `stick=0` candidate win.

**2. `insert_restickify.py`** — applies the decision. An internal buffer has no caller-owned address and no pre-existing data, so unlike a graph input it needs **no copy-back**:

```
insert_post_mutation_restickify             # runs after insert_restickify
└── for op in tagged_ops:                   # ops carrying _restickify_plan
    ├── graph input?  -> existing copy-back path (unchanged)
    │                    pre-restickify, copy-back, set_spyre_tensor_layout
    └── _retarget_internal_buf_mutation     # ADDED: replaces assert graph_input is not None
        └── rebind the write onto the alt-layout buffer via a fresh ReinterpretView
```

The internal buffer is *born* in the alt layout — propagation commits it, `finalize_layouts` wraps it, `spyre_empty_with_layout` allocates it — so the retarget only rebinds the write; there is nothing to repair afterwards and no `set_spyre_tensor_layout` call. Only the graph-input arm emits one, and graph I/O is never HBM-pooled, so it needs no pooling exception to survive. **This PR therefore leaves `hbm_pool_planning.py` untouched.**

A genuine multi-level view is rejected by assert as a loud tripwire rather than shipping untestable recomposition.

#### Which issue(s) this PR is related to:

Fixes #1464 (Identity handling for values smaller than a stick along the stick dimension) — the follow-up promised in https://github.com/torch-spyre/torch-spyre/issues/1464#issuecomment-4693069453: #2595 enabled unaligned slices on *input* tensors, and this PR does the same for *output* (mutation) buffers. It is also step 2 of the plan in https://github.com/torch-spyre/torch-spyre/issues/1464#issuecomment-4507833586; step 3 (#1756) has since landed, so nothing else blocks #1464.

Related to #1094 (Enable/test torch.cat for Q2 & Q3 target models).

Scope note on #1464: it names `torch.stack` and `torch.nn.functional.pad` alongside `torch.cat`. The mutation-buffer mechanism fixed here is the one all three share, but only `torch.cat` is covered by tests — `stack` and `pad` are unverified and may need their own lowering work.

#### Special notes for your reviewer:

**One shape now raises instead of miscompiling.** An offset-free (`start=0`) sub-stick write into a **graph input** is rejected in `propagate_layouts` with a named `Unsupported`, before any alt layout is committed — preserving the stick tail across a copy-back is not representable on that path. On `main` this shape silently returns wrong results, so the raise is a strict improvement. The same write into an **internal** buffer is supported and goes through the relocation above.

**Known gap, deferred to a follow-up: a size-1 stick dim.** When the stick dim has size 1, Inductor collapses its range-1 loop and the stick access arrives as a bare integer constant with no loop variable. `is_stick_expr_offset_free` admits only `Mod(var, 64)`, a bare symbol, or literal `0`, so a nonzero constant offset (`[..., 3:4]`) matches no admitted form and raises. This is pre-existing, affects reads and writes through that one shared predicate, and no shape here reaches it — `torch.cat` on a size-1 stick dim is the case to watch.

**Test coverage.** `test_strict_cat_model_shapes` covers the target-model shapes: the four aligned and non-stick-dim shapes take the no-relocation path, and the two mid-stick 32+32 shapes relocate both writes onto one shared layout. The `slice_scatter` param sets `2d_substick_off32` and `substick_off32_3d` cover the offset sub-stick write on the compiled-op path, over a stick dim spanning four sticks and a single stick respectively.

**Not covered, for the record.** Neither `raise Unsupported` path has a test. No test produces an internal-buffer group whose writers want *different* alts, so the "first alt wins, force the rest" resolution is exercised only where all writers independently agree (measured: two multi-writer groups, both agreeing); its safety under genuine disagreement rests on the retarget rebinding every writer onto one buffer, not on test evidence.

#### Does this PR introduce a user-facing change?

Yes. `torch.cat` whose split point is not a stick multiple (e.g. the mid-stick 32+32 rotary-embedding concats) previously returned wrong results silently, with a zero exit status; it now compiles correctly. More generally, offset and sub-stick slice writes into compiler-allocated buffers no longer miscompile. One shape that previously miscompiled silently — an offset-free sub-stick write into a graph input — now raises `Unsupported` instead.

### CODE REVIEW

**Comment by yoheiueda** [comment]
@tardieu @dgrove-oss @marnold-ibm 
This PR adds support for torch.cat. Could you please take a look?

File: torch_spyre/_inductor/propagate_layouts.py [hunk 1]
```diff
@@ -1606,15 +1609,169 @@ def _target_device_layout(target, name: str):
...
+    A ``[..., :32]`` slice into a 64-wide stick is offset-free yet shorter than
+    the stick, and writing it natively zeroes the tail. Requires a single loop
+    var whose extent is strictly between 0 and ``stick_size`` over a full-stick
+    dim, so a dim naturally narrower than one stick is left alone.
+    """
+    syms = write_stick.free_symbols
+    if len(syms) != 1:
+        return False
+    var = next(iter(syms))
+    extent = getattr(output_dep, "ranges", {}).get(var)
```
Comment (maxdebayser) [inline torch_spyre/_inductor/propagate_layouts.py]: MemoryDep.ranges is a property method. Is the gettarr really needed? It messes with python static code analysis.

File: torch_spyre/_inductor/propagate_layouts.py [hunk 2]
```diff
@@ -1639,6 +1799,32 @@ def _find_alt_target_stl(
             f"no offset-free alternative stick dim for mutation target "
             f"(write stick {write_stick!r}, size={target_layout.size})"
         )
+
+    # Rank by restickify feasibility: compute_restickify_target_layout returns
```
Comment (maxdebayser) [inline torch_spyre/_inductor/propagate_layouts.py]: It's not really a ranking, the loop below returns on the first feasible layout it encounters without ranking all other layouts.

**Comment by maxdebayser** [comment]
I've tested several (binary) combinations of input shapes and views on input shapes and they all generate the correct result.

But I've also found some opportunities for optimization. The program below with the example inputs generates 9 opspecs, of which the majority are identity ops, in other words, copies. 
```
a2 = torch.full((2,32), 1, device="spyre", dtype=torch.float16)
b2 = torch.full((2,32), 2, device="spyre", dtype=torch.float16)
c2 = torch.full((2,64), 3, device="spyre", dtype=torch.float16)

# original program
@torch.compile
def forward(a,b,c, dim=-1):
   return torch.cat((a,b), dim=dim) * c
```

Some of these can be fused with the restickify operations. By editing the opspecs directly, I was able to get the number down to 5. But maybe this is better suited for future work to more comprehensively fuse restickifies with identity ops.

File: torch_spyre/_inductor/propagate_layouts.py (hunk 1 above)
Comment (yoheiueda) [reply torch_spyre/_inductor/propagate_layouts.py]: Done

File: torch_spyre/_inductor/propagate_layouts.py (hunk 2 above)
Comment (yoheiueda) [reply torch_spyre/_inductor/propagate_layouts.py]: Comment updated

**Comment by yoheiueda** [comment]
@maxdebayser Thanks for the review!

I have addressed the other comments; the opspec count is the one that needed a longer answer. I measured what the compiler actually emits for your case, and the breakdown turned out different from "mostly redundant copies", so it seems worth recording before anyone attempts the fusion.

Naming, so the table reads: your case is `cat((a, b), dim=-1) * c` with `a`, `b` of shape `(2, 32)` and `c` of shape `(2, 64)`, all fp16. `arg0_1`, `arg1_1`, `arg2_1` are those three graph inputs in order, `arg_4` is the output. Device size is the tiled allocation each op addresses; the **bold** dim on rows 1-3 is the one being padded, `2 → 64`. Back gap is what an op leaves unwritten in its own output.

| # | op | in | out | writes at | dev size | SDSC |
|---|---|---|---|---|---|---|
| 1 | identity | arg0_1 | buf7 | [0, c0, c1] | in: [1, **2**, 64]<br>out: [1, **64**, 64] | iteration: {mb:2, out:64}<br>back gap: {mb 62} |
| 2 | identity | arg1_1 | buf8 | [0, c0, c1] | in: [1, **2**, 64]<br>out: [1, **64**, 64] | iteration: {mb:2, out:64}<br>back gap: {mb 62} |
| 3 | identity | arg2_1 | buf9 | [0, c0, c1] | in: [1, **2**, 64]<br>out: [1, **64**, 64] | iteration: {mb:2, out:64}<br>back gap: {mb 62} |
| 4 | ReStickify | buf7 | buf4 | [0, c1, c0] | in: [1, 64, 64]<br>out: [1, 64, 64] | iteration: {mb:64, out:64}<br>back gap: {} |
| 5 | ReStickify | buf8 | buf5 | [0, c1, c0] | in: [1, 64, 64]<br>out: [1, 64, 64] | iteration: {mb:64, out:64}<br>back gap: {} |
| 6 | ReStickify | buf9 | buf6 | [0, c1, c0] | in: [1, 64, 64]<br>out: [1, 64, 64] | iteration: {mb:64, out:64}<br>back gap: {} |
| 7 | identity | buf4 | buf0 | [0, c1, c0] | in: [1, 64, 64]<br>out: [1, 64, 64] | iteration: {mb:64, out:32}<br>back gap: {out 32} |
| 8 | identity | buf5 | buf0 | [0, c1 + 32, c0] | in: [1, 64, 64]<br>out: [1, 64, 64] | iteration: {mb:64, out:32}<br>back gap: {out 32} |
| 9 | mul | buf0, buf6 | arg_4 | [0, c1, c0] | in: [1, 64, 64]<br>out: [1, 64, 64] | iteration: {mb:64, out:64}<br>back gap: {} |

I have grouped by stage (three pads, three transposes, two concat writes, the multiply) so each input's chain reads down the table. Note that rows 3 and 6 are the multiplier `c`, not a third `cat` operand: it needs the same transpose to match the concatenated operands' orientation.

Reconciling against your 5, if I have reconstructed the hand-edit correctly:

```
restickify   arg0_1 → buf0 [0,c1,c0]      transpose a, write it at offset 0   (rows 4 + 7)
restickify   arg1_1 → buf5                transpose b into a temp             (row 5)
restickify   arg2_1 → buf6                transpose c                         (row 6)
identity     buf5   → buf0 [0,c1+32,c0]   copy the temp to offset 32          (row 8)
mul          buf0, buf6 → arg_4                                               (row 9)
```

Grouped the same way as the table, so the correspondence is visible: the three restickifies read the graph inputs directly, rows 1-3 are gone, and row 7 has been folded into the first restickify while row 8 survives as a copy.

**Rows 1-3 are not redundant copies. They are what makes the restickify's read legal.** The restickify instruction always reads its padded non-stick dim in full (the dim that becomes the stick dim afterwards) and always writes the padded non-stick dim in full (the one that was the stick dim). Both are hardware constraints, and "in full" means to the stick boundary: 64 rows here, since the dim being padded is the size-2 `mb` dim, not the host width.

But `arg0_1`'s allocation is 2 rows. A restickify reading it directly would run 62 rows past the caller's tensor, and there is no narrower version of the read. Something has to produce a 64-row source first, which is what rows 1-3 do. When the source is an intermediate the producer's output is padded in place and no copy appears; a **graph input** cannot be widened, since the allocation is the caller's tensor, so it is cloned into a padded buffer instead. That is #3110's documented behaviour, and it is also why these three cannot be shared: they are three distinct graph inputs. So each of your three restickifies needs one clone in front, which takes 5 back to 8.

- #3110

The back-gap values in the SDSC column show this from the other side. Rows 1-3 write 2 of the 64 rows they allocate, and the 62 they skip are exactly the padding that makes the next op's full-width read legal. Rows 4-6 have no gap at all, which is not the absence of a constraint but the signature of one, since a restickify reads 64 and writes 64 and so has nothing left over to record.

**Rows 7 and 8 are the genuinely reducible pair, and that part is a real opportunity.** Each is a pure copy from a restickify's temp into a slice of `buf0` (`lower_cat` emits one `mutate_to` per operand, [`lowering.py:1568`](https://github.com/torch-spyre/torch-spyre/blob/8a0df1c11470ebb17f0903779cd604fe6f1c8733/torch_spyre/_inductor/lowering.py#L1568)), so fusing means letting the restickify write the operand's final coordinate directly, `[0, c1, c0]` and `[0, c1 + 32, c0]`, instead of filling a temp that is then copied. This is #2278: the same compute-into-temp-then-copy-into-a-slice pattern, and what it proposes is exactly your hand-edit. Both operands are the same case, so one pass covers both offsets.

- #2278 

It needs a new compiler pass, so I think #2278 should be its own PR rather than part of this one. What do you think?

**Comment by maxdebayser** [comment]
Hi @yoheiueda, I thought it was ok for restickify to read a bit beyond the tensor boundaries because in the destination buffer this extra data would be overwritten anyway, hence the opportunity to remove some copies. But Olivier explained to me that this is a big no-no. Regarding the other copy elisions that you verified, I think it makes perfect sense to do that in another PR focused specifically on that.
With the caveat that my understanding of this codebase is still incomplete, I think the changes in this PR make sense. As I mentioned before, I also ran the code locally with several different shapes and couldn't find any problems.

**Review by maxdebayser — APPROVED** [review]

**Comment by yoheiueda** [comment]
@maxdebayser Thank you for the comment!


> I thought it was ok for restickify to read a bit beyond the tensor boundaries because in the destination buffer this extra data would be overwritten anyway, hence the opportunity to remove some copies.

My understanding is that this is right on your shapes, and not on others, so it may be worth writing down. As far as I can tell a device allocation has no gaps, so each dim's stride is the product of the `device_size` entries inside it. Padding grows exactly one of those entries, which changes the stride of every dim outside the padded one and leaves the rest alone. Whether that actually moves an element then depends on whether any of those outer dims has more than one element, since a size-1 dim's coordinate is always 0.

In your example there is only one dim outside the padded one and it has size 1: `device_size` is `[1, 2, 64]` and the padded dim is the 2. So growing it to 64 should move nothing, and the padded layout is the original declared longer. The extra data lands in the output's padding columns, which would explain why the result comes out correct and only the read is illegal.

Once a dim outside the padded one is real, as in a 3D or 4D input, I think its stride is a product that includes the padded extent, so bumping 8 to 64 moves it from 8 * 64 to 64 * 64. If that is right, the padded layout becomes a different address function rather than a longer one: the valid rows would have to be moved, not merely followed by more bytes. No read of the original input can produce that, so there the copy seems to be doing real work.

**Comment by yoheiueda** [comment]
In an offline discussion, @avery-blanchard  raised a concern about the HBM planning implications of the `_emit_set_layout` changes in this PR.

After investigating, I found that the `_emit_set_layout` tag is redundant for the internal buffer, since it is allocated directly in alt_stl. 

As a result, I was able to remove both the tag and the associated HBM planning changes. I also included some related cleanup.

**Review by avery-blanchard — APPROVED** [review]
looks good to me, thanks!
@marnold-ibm could you take a quick look also?

**Comment by marnold-ibm** [comment]
@avery-blanchard @yoheiueda looks good to me too.

I'll click merge!
